src/Storages/MergeTree/GinIndexStore.cpp

```diff
@@ -35,123 +35,113 @@ const CompressionCodecPtr & GinIndexCompressionFactory::zstdCodec()
 {
-    if (containsAllRows())
-        return;
+    rowid_bitmap.add(row_id);
+}
 
-    if (useRoaring())
-    {
-        if (rowid_bitmap.cardinality() == size_limit)
-        {
-            /// reset the postings list with MATCH ALWAYS;
-            rowid_lst_length = 1; /// makes sure useRoaring() returns false;
-            rowid_lst[0] = CONTAINS_ALL; /// set CONTAINS_ALL flag;
-        }
-        else
-            rowid_bitmap.add(row_id);
-    }
-    else
+UInt64 GinIndexPostingsBuilder::serialize(WriteBuffer & buffer)
+{
+    rowid_bitmap.runOptimize();
+
+    const UInt64 cardinality = rowid_bitmap.cardinality();
+
+    if (cardinality < MIN_SIZE_FOR_ROARING_ENCODING)
     {
-        assert(rowid_lst_length < MIN_SIZE_FOR_ROARING_ENCODING);
-        rowid_lst[rowid_lst_length] = row_id;
-        rowid_lst_length++;
+        std::vector<UInt32> values(cardinality);
+        rowid_bitmap.toUint32Array(values.data());
 
-        if (rowid_lst_length == MIN_SIZE_FOR_ROARING_ENCODING)
-        {
-            for (size_t i = 0; i < rowid_lst_length; i++)
-                rowid_bitmap.add(rowid_lst[i]);
+        UInt64 header = (cardinality << 1) | 0x1; /// Use array
+        writeVarUInt(header, buffer);
 
-            rowid_lst_length = USES_BIT_MAP;
+        UInt64 written_bytes = getLengthOfVarUInt(header);
+        for (const auto & value : values)
+        {
+            writeVarUInt(value, buffer);
+            written_bytes += getLengthOfVarUInt(value);
         }
+
+        return written_bytes;
     }
-}
 
-UInt64 GinIndexPostingsBuilder::serialize(WriteBuffer & buffer)
-{
-    UInt64 written_bytes = 0;
-    buffer.write(rowid_lst_length);
-    written_bytes += 1;
+    const bool compress = cardinality >= ROARING_ENCODING_COMPRESSION_CARDINALITY_THRESHOLD;
+    const UInt64 uncompressed_size = rowid_bitmap.getSizeInBytes();
 
-    if (useRoaring())
-    {
-        rowid_bitmap.runOptimize();
-        auto size = rowid_bitmap.getSizeInBytes();
-        auto buf = std::make_unique<char[]>(size);
-        rowid_bitmap.write(buf.get());
+    auto buf = std::make_unique<char[]>(uncompressed_size);
+    rowid_bitmap.write(buf.get());
 
-        const auto & codec = GinIndexCompressionFactory::zstdCodec();
+    UInt64 header = (uncompressed_size << 1);
+    header <<= 1;
+    if (compress)
+    {
         Memory<> memory;
-        memory.resize(codec->getCompressedReserveSize(static_cast<UInt32>(size)));
-        auto compressed_size = codec->compress(buf.get(), static_cast<UInt32>(size), memory.data());
+        const auto & codec = GinIndexCompressionFactory::zstdCodec();
+        memory.resize(codec->getCompressedReserveSize(static_cast<UInt32>(uncompressed_size)));
+        auto compressed_size = codec->compress(buf.get(), static_cast<UInt32>(uncompressed_size), memory.data());
 
-        writeVarUInt(size, buffer);
-        written_bytes += getLengthOfVarUInt(size);
+        header |= 0x2; /// zstd compressed + use roaring
```

Comment:
Guess it would be nice to store the on-disk compression types as separate int constants somewhere which we can reference in `serialize` and `deserialize`.